File: frontend/src/lib/whip.ts
const GATEWAY_URL = import.meta.env.VITE_GATEWAY_URL ?? ''

// postWhipOffer performs the WHIP bootstrap: it POSTs the local SDP offer as
// `application/sdp` and returns the SFU's answer SDP. Room and member are
// passed as query parameters since the body is pure SDP.
export async function postWhipOffer(
  roomId: string,
  memberId: string,
  offerSdp: string,
): Promise<string> {
  const url = `${GATEWAY_URL}/whip?room=${encodeURIComponent(roomId)}&member=${encodeURIComponent(memberId)}`

  const res = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/sdp' },
    body: offerSdp,
  })

  if (!res.ok) {
    const detail = await res.text().catch(() => '')
    throw new Error(`WHIP request failed (${res.status}): ${detail || res.statusText}`)
  }

  return res.text()
}

File: internal/gateway/rest/whip.go
package rest

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"time"
)

const (
	// maxOfferSize caps the SDP offer body the gateway is willing to read.
	maxOfferSize = 256 << 10 // 256 KiB

	// connectTimeout bounds how long the gateway waits on the streaming service.
	connectTimeout = 15 * time.Second
)

// handleWHIP implements WebRTC-HTTP Ingestion Protocol bootstrap signaling.
//
// The publisher POSTs its SDP offer as `application/sdp`, identifying the room
// and member via query parameters (`?room=<id>&member=<id>`). The gateway
// forwards the offer to the streaming SFU and returns its answer SDP. All
// further signaling (renegotiation as peers join, leave notifications) flows
// over the peer's WebRTC DataChannel.
func (h *Handler) handleWHIP(w http.ResponseWriter, r *http.Request) {
	roomID := r.URL.Query().Get("room")
	memberID := r.URL.Query().Get("member")
	if roomID == "" || memberID == "" {
		http.Error(w, "missing room or member query parameter", http.StatusBadRequest)
		return
	}

	if ct := r.Header.Get("Content-Type"); ct != "" && ct != "application/sdp" {
		http.Error(w, "expected Content-Type: application/sdp", http.StatusUnsupportedMediaType)
		return
	}

	offer, err := io.ReadAll(io.LimitReader(r.Body, maxOfferSize))
	if err != nil {
		http.Error(w, "failed to read offer", http.StatusBadRequest)
		return
	}
	if len(offer) == 0 {
		http.Error(w, "empty offer", http.StatusBadRequest)
		return
	}

	ctx, cancel := context.WithTimeout(r.Context(), connectTimeout)
	defer cancel()

	answer, err := h.signaling.Connect(ctx, roomID, memberID, string(offer))
	if err != nil {
		h.logger.Error("WHIP connect failed",
			slog.String("room", roomID),
			slog.String("member", memberID),
			slog.String("error", err.Error()),
		)
		http.Error(w, "failed to establish session", http.StatusBadGateway)
		return
	}

	w.Header().Set("Content-Type", "application/sdp")
	w.Header().Set("Location", fmt.Sprintf("/whip?room=%s&member=%s",
		url.QueryEscape(roomID), url.QueryEscape(memberID)))
	w.WriteHeader(http.StatusCreated)

	if _, err := w.Write([]byte(answer)); err != nil {
		h.logger.Error("WHIP write answer failed", slog.String("error", err.Error()))
	}
}
